dolphinscheduler-dao/src/main/java/org/apache/dolphinscheduler/dao/repository/impl/TaskInstanceContextDaoImpl.java

```diff
@@ -0,0 +1,106 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.dolphinscheduler.dao.repository.impl;
+
+import org.apache.dolphinscheduler.common.constants.Constants;
+import org.apache.dolphinscheduler.common.enums.ContextType;
+import org.apache.dolphinscheduler.common.utils.JSONUtils;
+import org.apache.dolphinscheduler.dao.entity.AbstractTaskInstanceContext;
+import org.apache.dolphinscheduler.dao.entity.DependentResultTaskInstanceContext;
+import org.apache.dolphinscheduler.dao.entity.TaskInstanceContext;
+import org.apache.dolphinscheduler.dao.mapper.TaskInstanceContextMapper;
+import org.apache.dolphinscheduler.dao.repository.BaseDao;
+import org.apache.dolphinscheduler.dao.repository.TaskInstanceContextDao;
+
+import org.apache.commons.collections4.CollectionUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.TreeSet;
+import java.util.stream.Collectors;
+
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class TaskInstanceContextDaoImpl extends BaseDao<TaskInstanceContext, TaskInstanceContextMapper>
+        implements
+            TaskInstanceContextDao {
+
+    public TaskInstanceContextDaoImpl(TaskInstanceContextMapper taskInstanceContextMapper) {
+        super(taskInstanceContextMapper);
+    }
+
+    @Override
+    public List<TaskInstanceContext> queryListByTaskInstanceIdAndContextType(Integer taskInstanceId,
+                                                                             ContextType contextType) {
+        if (taskInstanceId == null) {
+            return Collections.emptyList();
+        }
+        return mybatisMapper.queryListByTaskInstanceIdAndContextType(taskInstanceId, contextType);
+    }
+
+    @Override
+    public int deleteByTaskInstanceIdAndContextType(Integer taskInstanceId, ContextType contextType) {
+        if (taskInstanceId == null) {
+            return 0;
+        }
+        return mybatisMapper.deleteByTaskInstanceIdAndContextType(taskInstanceId, contextType);
+    }
+
+    @Override
+    public int upsertTaskInstanceContext(TaskInstanceContext taskInstanceContext) {
+        if (taskInstanceContext == null) {
+            return 0;
+        }
+        TaskInstanceContext dbTaskInstanceContext =
+                mybatisMapper.queryListByTaskInstanceIdAndContextType(taskInstanceContext.getTaskInstanceId(),
+                        taskInstanceContext.getContextType()).stream().findFirst().orElse(null);
```

Comment:
One task instance id with a specific context type should only contain one record.